### PR TITLE
SI-9064: Add reduceByKey to Iterable

### PR DESCRIPTION
**Pull request by angellandros**
This is the first function of a series of useful functions. I just wanted to check whether the community likes the whole idea or not. If this idea is acceptable, I will add other functions also.

### CODE REVIEW

**Comment by kanielc** [comment]
You should probably go through the mailing list first.
https://groups.google.com/forum/#!forum/scala-internals

Also, these new functions would need to be based on 2.12.x instead of 2.11, as they would break forward compatibility.

**Comment by angellandros** [comment]
Thank you, Denton. I started a topic:
https://groups.google.com/forum/#!topic/scala-internals/fKx_kN3Op4A
And I will port my changes to the newer branch.

File: src/library/scala/collection/Iterable.scala [hunk 1]
```diff
@@ -52,3 +57,19 @@ object Iterable extends TraversableFactory[Iterable] {
 
 /** Explicit instantiation of the `Iterable` trait to reduce class file size in subclasses. */
 abstract class AbstractIterable[+A] extends AbstractTraversable[A] with Iterable[A]
+
+/** PairedIterable for functions on paired iterables */
+class PairedIterable[K, V](x: Iterable[(K, V)]) {
+  def reduceByKey(func: (V,V) => V) = {
```
Comment (Ichoran) [inline src/library/scala/collection/Iterable.scala:63]: Isn't this a special case of a method `reduceToMap[K, V](keyFn: A => K, valFn: A => V, merge: (V, V) => V)`?  Or shouldn't it instead be defined on TraversableOnce so you can do the key/value mapping on the fly (e.g. `xs.iterator.map(a => (k(a), v(a))).reduceByKey(foo)`)?

Comment (angellandros) [inline src/library/scala/collection/Iterable.scala:63]: Never heard of `reduceToMap`. And couldn't find it. But it seems to have a map, and then a reduce. I am just having the reduce here. And, you are right. I should add it there, but I had some technical difficulties to do so. Anyway, I can do that.
`l.map(x => (x,1)).reduceByKey(_+_)`
Take a look at this:
http://ideone.com/dyrkYM

**Comment by angellandros** [comment]
I had an optimization for my implementation. And, as Rex said in the topic in scala internals, policy is not to have new features until the next version. So, I am closing the request.

**Pull request closed without merging**